fdbclient/NativeAPI.actor.cpp

```diff
@@ -5267,6 +5347,7 @@ ACTOR static Future<Void> tryCommit(Reference<TransactionState> trState,
 					if (CLIENT_BUGGIFY) {
 						throw commit_unknown_result();
```

Comment:
This `commit_unknown_result` reminds me: it's kind of important (though not _strictly_ necessary for correctness) that clients using cached read versions don't use that cached version in retry loops on attempts after the first one. For example, you could imagine that if the staleness bound is generous enough, that a transaction that fails due to a conflict (and therefore doesn't update the cached read version after their failed commit), might get the same read version on a second attempt as they did the first attempt, which will then cause them to hit the same conflict as the first time.

In the Record Layer, we dealt with this by always clearing out the `WeakReadSemantics` within our retry loops if it's not the initial attempt: https://github.com/FoundationDB/fdb-record-layer/blob/5efeb2355be9507a9802d18439e53016dfc1cac3/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunnerImpl.java#L172 That way, we would ask for a new read version from the proxy if it's a retry. I'm not sure if there's some clever way of making that automatically happen at the FDB client level (e.g., by having `onError` set some state in the new transaction indicating that it should skip GRV caching even if requested), or if the retry logic in the language bindings should be updated.

Reply:
I think this is actually taken care of internally in the error `catch` block of the commit path: https://github.com/apple/foundationdb/blob/main/fdbclient/NativeAPI.actor.cpp#L5353 `commitDummyTransaction` ensures that a newer read version will be committed and cached, which should be able to avoid this conflict. This is also lightly tested in the workload I added, `SidebandSingle`. The exception is if the transaction is using `causalWriteRisky`, but I think any expectation of consistency should be discarded if that option is being used anyways.

Reply:
`causalWriteRisky` is currently documented as allowing a transaction to commit twice (and I think what this actually means is that your transaction could commit at some point in the future). We probably wouldn't want using this option to exhibit the behavior that Alec described, but I think the fact that this only comes into play when we get `commit_unknown_result` or `request_maybe_delivered` means that we shouldn't have to worry about retries repeatedly running into the same error.

However, I do wonder if other errors pose a problem in general. `commitDummyTransaction` is not called if you get a `not_committed` error, for example, so Alec's example of running into a conflict repeatedly could still apply if you don't have some other way to update the cached read version or bypass the cache.